Yafc/Widgets/DataGrid.cs

```diff
@@ -1,23 +1,69 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Numerics;
+using System.Reflection;
 using SDL2;
 
 namespace Yafc.UI {
-    public abstract class DataColumn<TData>(float width, float minWidth = 0f, float maxWidth = 0f) {
-        public readonly float minWidth = minWidth == 0f ? width : minWidth;
-        public readonly float maxWidth = maxWidth == 0f ? width : maxWidth;
+    public abstract class DataColumn<TData> {
+        public readonly float minWidth;
+        public readonly float maxWidth;
         public readonly bool isFixedSize = minWidth == maxWidth;
-        public float width = width;
+        private readonly Func<float> getWidth;
+        private readonly Action<float> setWidth;
+        private float _width;
+
+        /// <param name="widthStorage">If not <see langword="null"/>, names an instance property in <see cref="Preferences"/> that will be used to store the width of this column.
+        /// If the current value of the property is out of range, the initial width will be <paramref name="initialWidth"/>.</param>
+        public DataColumn(float initialWidth, float minWidth = 0f, float maxWidth = 0f, string? widthStorage = null) {
+            this.minWidth = minWidth == 0f ? initialWidth : minWidth;
+            this.maxWidth = maxWidth == 0f ? initialWidth : maxWidth;
+            if (widthStorage != null) {
+                (getWidth, setWidth) = getStorage(widthStorage);
+            }
+            else {
+                getWidth = () => _width;
+                setWidth = f => _width = f;
+            }
+
+            if (width < this.minWidth || width > this.maxWidth) {
+                width = initialWidth;
+            }
+
+            static (Func<float>, Action<float>) getStorage(string storage) {
+                Exception? innerException = null;
+                try {
+                    PropertyInfo? property = typeof(Preferences).GetProperty(storage);
+                    Func<float>? getMethod = property?.GetGetMethod()?.CreateDelegate<Func<float>>(Preferences.Instance);
+                    Action<float>? setMethod = property?.GetSetMethod()?.CreateDelegate<Action<float>>(Preferences.Instance);
+                    if (getMethod == null || setMethod == null) {
+                        goto @throw;
```

Comment:
I am not well-versed in exception handling, so I wonder if throwing here, then catching with `catch` and re-throwing would be easier to follow than `goto` in two places.

Reply:
The/This `goto` is basically breaking out of the exception handler below. The other goto is basically reusing the 'throw exception' code. Yes you could throw an exception and rethrow again in the `catch`-er.

Whether this improved readability or not, I have my doubts.

Reply:
I wanted to avoid string duplication in the source. On further thought I decided to change the exceptions to only mention errors that could cause that exception. (e.g. a non-float property cannot cause the first exception, so I don't mention that there.)